wave_utils/include/wave/utils/config.hpp

```diff
@@ -25,57 +25,54 @@ namespace wave {
 /** @addtogroup utils
  *  @{ */
 
-/**
- * An enum used by `ConfigParam` to denote the yaml value type.
- *
- *  Currently we support parsing of the following data types:
- *  - **Primitives**: `bool`, `int`, `float`, `double`, `std::string`
- *  - **Arrays**: `std::vector<bool>`, `std::vector<int>`, `std::vector<float>`,
- * `std::vector<double>`, `std::vector<std::string>`
- *  - **Vectors**: `Eigen::Vector2d`, `Eigen::Vector3d`, `Eigen::Vector4d`,
- * `Eigen::VectorXd`
- *  - **Matrices**: `Eigen::Matrix2d`, `Eigen::Matrix3d`, `Eigen::Matrix4d`,
- * `Eigen::MatrixXd`
- */
-enum ConfigDataType {
-    TYPE_NOT_SET = 0,
-    // PRIMITIVES
-    BOOL = 1,
-    INT = 2,
-    FLOAT = 3,
-    DOUBLE = 4,
-    STRING = 5,
-    // ARRAY
-    BOOL_ARRAY = 11,
-    INT_ARRAY = 12,
-    FLOAT_ARRAY = 13,
-    DOUBLE_ARRAY = 14,
-    STRING_ARRAY = 15,
-    // VECTOR
-    VEC2 = 21,
-    VEC3 = 22,
-    VEC4 = 23,
-    VECX = 24,
-    // MATRIX
-    MAT2 = 31,
-    MAT3 = 32,
-    MAT4 = 33,
-    MATX = 34,
-    CVMAT = 35
+/** Return codes for ConfigParser methods */
+enum class ConfigStatus {
+    OK = 0,                  ///< Success
+    MissingOptionalKey = 1,  ///< key not found, but parameter is optional
+    // Note: errors are != ConfigStatus::OK
+    FileError = -1,      ///< Success
+    KeyError = -2,       ///< key not found, parameter is not optional
+    ConversionError = -3 /**< Invalid value for conversion
+                           * e.g. wrong-size list for vector, missing
+                           * key 'rows', 'cols' or 'data' (for matrix) */
+};
+
+/** Base class representing a parameter to be parsed in the yaml file */
+struct ConfigParamBase {
+    ConfigParamBase(std::string key, bool optional)
+        : key{std::move(key)}, optional{optional} {};
```

Comment:
why move? typo?

Reply:
See https://stackoverflow.com/questions/7592630/is-pass-by-value-a-reasonable-default-in-c11
https://clang.llvm.org/extra/clang-tidy/checks/modernize-pass-by-value.html

Reply:
This just hurts my head.